presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -62,22 +62,42 @@ private ExpressionTreeFilter() {}
         });
     }
 
-    public static List<Expression> extract(Function<Node, Optional<Expression>> predicate, List<? extends Node> nodes)
+    public static <T extends Expression> List<T> extractExpressionsOfTypeUsingPredicate(
+            Iterable<? extends Node> nodes,
+            Class<T> clazz,
+            Function<Node, Optional<T>> predicate)
     {
-        requireNonNull(predicate, "predicate is null");
         requireNonNull(nodes, "nodes is null");
+        requireNonNull(clazz, "clazz is null");
+        requireNonNull(predicate, "predicate is null");
 
-        ImmutableList.Builder<Expression> extractedExpressions = ImmutableList.builder();
+        ImmutableList.Builder<T> expressionsBuilder = ImmutableList.builder();
         for (Node node : nodes) {
             List<? extends Node> flattenedNodes = linearizeNodes(node);
-            extractedExpressions.addAll(flattenedNodes.stream()
+            expressionsBuilder.addAll(flattenedNodes.stream()
                     .map(predicate)
                     .filter(Optional::isPresent)
                     .map(Optional::get)
+                    .map(clazz::cast)
                     .collect(toImmutableList()));
         }
 
-        return extractedExpressions.build();
+        return expressionsBuilder.build();
+    }
+
+    public static boolean hasExpressionOfType(List<Expression> expressions, Class clazz)
+    {
+        requireNonNull(expressions, "expressions is null");
```

Comment:
I think having a separate method is more readable but I don't really feel strongly so I'll implement your suggestion.